File: nvpl_blas/f77/dscal.f
!******************************************************************************
! Content:
!     This example demonstrates use of API as below:
!     DSCAL 
!
!******************************************************************************
      program   DSCAL_MAIN
*
      implicit none

      include "nvpl_blas_f77_blas.fi"

*       Intrinsic Functions
      intrinsic        abs

      integer          n, incx, i
      parameter        (n=5, incx=1)
      double precision alpha
      parameter        (alpha=1.0)
      integer          xsize
      parameter        (xsize=1+(n-1)*abs(incx))
      double precision x(xsize)
      data             (x(i), i=1,xsize)/1.0,-3.7,4.3,0.0,5.0/

*       External Subroutines
      external         print_dvector 
*
*      Executable Statements
*
      print*
      print 99
*
*       Print input data
      print*
      print 100, n, incx, alpha
      call print_dvector(x,n,incx,'X')
*
*      Call DSCAL subroutine
      call DSCAL(n,alpha,x,incx)
*
      print*
      print 101
      call print_dvector(x,n,incx,'X')

      stop
 99   format('Example: DSCAL for multiplying the vector x by the'
     &       ' scalar')
 100  format('#### args: n=',i1,', incx=',i1,', alpha=',f4.2)
 101  format('The result of DSCAL: ')
      end

File: nvpl_blas/c/dscal.c
/******************************************************************************
 * Content: 
 *     This example demonstrates use of API as below:
 *     cblas_dscal
 *
 ******************************************************************************/
#include "example_helper.h"

int main() {
    nvpl_int_t N = 5;
    double * X;
    double alpha = 1.0f;
    nvpl_int_t incX = 1;

    printf("\nExample: cblas_dscal for multiplying the vector x by the scalar\n");
    printf("#### args: n=%" PRId64 ", incx=%" PRId64 ", alpha=%f\n", (int64_t)N,
            (int64_t)incX, alpha);

    nvpl_int_t len_x = 1 + (N - 1) * labs(incX);
    // allocate memory
    X = (double *)malloc(len_x * sizeof(double));

    // fill data
    fill_dvector(X, N, incX);

    // print input data
    print_dvector(X, N, incX, "X");

    // call cblas_dscal
    cblas_dscal(N, alpha, X, incX);

    // print result
    printf("\nThe result of cblas_dscal:\n");
    print_dvector(X, N, incX, "X");

    // release memory
    free(X);
    return 0;
}
